### PR TITLE
[WIP] feat(organisms): add SfMegaMenu component

### LINKED ISSUE #18: [FEAT] MegaMenu component
Create mega menu component (desktop only)
This component is only a wrapper for other components so the task does not involve creating  the components inside of it.

image

Proposed API:
- just a default slot

Visibility properties shouldn't be a part of the component itself so the task is to make static MegaMenu component without any content, just a slot + write basic unit tests:
- if it renders
- if the slot content is rendered.

Designs: https://www.figma.com/file/hrwE3VsMBHgdJoS86rVr4W/Desktop-%26-Mobile-Vue-Storefront?node-id=99%3A0

### PR DESCRIPTION
**Pull request by aniamusial**
# Related issue
Closes #18

# Scope of work
edited by: @pspaczek  

### added
- [x] `SfMegaMenu and` _internal `SfMegaMenuColumn` components

### component API
```html
<SfMegaMenu parent-menu="Man">
  <SfMegaMenuColumn title="Man">...<SfMegaMenuColumn>
  <template #aside>
    ...
  </template>
</SfMegaMenu>
```
# Screenshots of visual changes
### SfMegaMenu
Zrzut ekranu 2019-12-10 o 21 46 45
Zrzut ekranu 2019-12-10 o 21 46 58

# Checklist

- [x] I followed [composition rules](https://docs.storefrontui.io/contributing/coding-guidelines.html#component-rules) for my component
- [x] I tested the component in most common device sizes (can be tested in Storybook [from viewport addon in top menu](https://github.com/storybooks/storybook/tree/master/addons/viewport))

### CODE REVIEW

File: packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.html [hunk 1]
```diff
@@ -1,4 +1,35 @@
-<section class="sf-mega-menu">
-  <!-- @slot -->
-  <slot/>
-</section>
+<div class="sf-mega-menu">
+    <section
```
Comment (przspa) [inline packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.html]:
```suggestion
    <div
```

File: packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.html [hunk 2]
```diff
@@ -1,4 +1,35 @@
...
+            </slot>
+          </SfList>
+        </SfList>
+      </div>
+      <div class="sf-mega-menu__sidebar-content">
+        <!-- @slot Slot for other content -->
+        <slot name="menu-content"/>
+      </div>
+    </section>
+    <section class="sf-mega-menu__custom-section">
```
Comment (przspa) [inline packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.html]:
```suggestion
    <div class="sf-mega-menu__custom-section">
```

File: packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.html [hunk 3]
```diff
@@ -1,4 +1,35 @@
...
+      <slot />
```
Comment (przspa) [inline packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.html]:
```suggestion
      <slot name="custom-section"/>
```

File: packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.html [hunk 4]
```diff
@@ -1,4 +1,35 @@
...
+                  :class="{'sf-mega-menu__menu-item--is-active': column === active}"
+                />
+              </span>
+            </slot>
+          </SfList>
+        </SfList>
+      </div>
+      <div class="sf-mega-menu__sidebar-content">
+        <!-- @slot Slot for other content -->
+        <slot name="menu-content"/>
```
Comment (przspa) [inline packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.html]:
```suggestion
        <slot/>
```
I think is better choice for default slot :)

File: packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.html [hunk 5]
```diff
@@ -1,4 +1,35 @@
...
+        <SfList class="sf-mega-menu__headers">
+          <SfList
+            v-for="column in columns"
+            :key="column"
+            class="sf-mega-menu__list-item">
+            <!-- @slot Custom menu-item markup -->
+            <slot
+              name="menu-item"
+              v-bind="{updateColumn, column, active}">
+              <span @click="updateColumn(column)">
```
Comment (przspa) [inline packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.html]: updateColumn? Probably it, not a descriptive name :)

File: packages/vue/src/components/organisms/SfMegaMenu/_internal/SfMegaMenuColumn.html [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<div v-if="isActive || !isMobile" class="sf-mega-menu-column">
```
Comment (przspa) [inline packages/vue/src/components/organisms/SfMegaMenu/_internal/SfMegaMenuColumn.html]:
```suggestion
<div class="sf-mega-menu-column">
```
Propably we don't need `v-if="isActive || !isMobile"`

File: packages/vue/src/components/organisms/SfMegaMenu/_internal/SfMegaMenuColumn.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      type: String,
+      default: ""
+    }
+  },
+  computed: {
+    isActive() {
+      return this.provided.active === this.title;
+    }
+  },
+  mounted() {
```
Comment (przspa) [inline packages/vue/src/components/organisms/SfMegaMenu/_internal/SfMegaMenuColumn.js]: move this behaviour to SfMegaMenu, like in ContentPages

File: packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.js [hunk 8]
```diff
@@ -1,3 +1,89 @@
...
+  name: "SfMegaMenu",
+  components: {
+    SfList,
+    SfMenuItem
+  },
+  props: {
+    /**
+     * Column title
+     */
+    title: {
```
Comment (przspa) [inline packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.js]:
```suggestion
```

File: packages/vue/src/components/organisms/SfMegaMenu/_internal/SfMegaMenuColumn.html (hunk 6 above)
Comment (aniamusial) [reply packages/vue/src/components/organisms/SfMegaMenu/_internal/SfMegaMenuColumn.html]: we do, otherwise other columns will be shown

**Comment by przspa** [comment]
@aniamusial please do CR :)

**Comment by filrak** [comment]
Something is wrong with this component 
image

File: packages/vue/src/components/organisms/SfFooter/SfFooter.js [hunk 9]
```diff
@@ -43,6 +43,10 @@ export default {
     }
   },
   methods: {
+    updateItems(title) {
```
Comment (filrak) [inline packages/vue/src/components/organisms/SfFooter/SfFooter.js]: what is the purpose of this method? does it come from some other PR?

File: packages/vue/src/components/organisms/SfHeader/SfHeader.md [hunk 10]
```diff
@@ -21,23 +26,25 @@ Header component for page navigation.
   <SfHeader
     title="Storefront UI"
     logo="/assets/logo.svg"
-    active-sidebar="activeSidebar"
+    active-icon="account"
   >
     <template #navigation>
-      <SfHeaderNavigationItem>WOMAN</SfHeaderNavigationItem>
-      <SfHeaderNavigationItem>MAN</SfHeaderNavigationItem>
+      <div class="sf-header__navigation-item">
```
Comment (filrak) [inline packages/vue/src/components/organisms/SfHeader/SfHeader.md]: Why you removed internal component?

File: packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.js [hunk 11]
```diff
@@ -1,4 +1,86 @@
...
-  name: "SfMegaMenu"
+  name: "SfMegaMenu",
+  components: {
+    SfList,
+    SfMenuItem,
+    SfBar
+  },
+  props: {
+    /**
+     * Parent menu
```
Comment (filrak) [inline packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.js]: not very descriptive, what exactly this prop is doing? ^^

File: packages/vue/src/components/organisms/SfMegaMenu/_internal/SfMegaMenuColumn.html [hunk 12]
```diff
@@ -0,0 +1,8 @@
+<div v-if="isActive" class="sf-mega-menu-column">
+    <!-- @slot Custom title markup -->
+    <slot name="title" v-bind="{ title }">
+      <h3 class="sf-mega-menu-column__title">{{ title }}</h3>
+    </slot>
+    <!-- @slot Custom column markup -->
+    <slot />
```
Comment (filrak) [inline packages/vue/src/components/organisms/SfMegaMenu/_internal/SfMegaMenuColumn.html]: Isn't this a default behavior to fill this slot? Why it's labeled as "custom"?

File: packages/vue/src/components/organisms/SfMegaMenu/_internal/SfMegaMenuColumn.js [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+     * Column title
+     */
+    title: {
+      type: String,
+      default: ""
+    }
+  },
+  computed: {
+    isActive() {
+      return this.$parent.active.includes(this.title);
```
Comment (filrak) [inline packages/vue/src/components/organisms/SfMegaMenu/_internal/SfMegaMenuColumn.js]: be careful with this.$parent.At least do some error handling

File: packages/vue/src/components/organisms/SfMegaMenu/_internal/SfMegaMenuColumn.js [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+    }
+  },
+  mounted() {
+    this.$parent.updateItems(this.title);
```
Comment (filrak) [inline packages/vue/src/components/organisms/SfMegaMenu/_internal/SfMegaMenuColumn.js]: Its acceptable to read from parent but modifying its data is anti-pattern. Please use provide/inject

File: packages/vue/src/examples/pages/layout/Layout.stories.js [hunk 15]
```diff
@@ -0,0 +1,14 @@
+/* eslint-disable import/no-extraneous-dependencies */
+import { storiesOf } from "@storybook/vue";
+import Layout from "./Layout";
```
Comment (filrak) [inline packages/vue/src/examples/pages/layout/Layout.stories.js]: Layout?

Comment (filrak) [reply packages/vue/src/examples/pages/layout/Layout.stories.js]: @pspaczek we already have stories for header and footer, you can just include this as a story  of megamenu (with header)

File: packages/vue/src/components/organisms/SfFooter/SfFooter.js (hunk 9 above)
Comment (pspaczek) [reply packages/vue/src/components/organisms/SfFooter/SfFooter.js]: it's a part of small SfFooter refactor, I remove it for this PR https://github.com/DivanteLtd/storefront-ui/pull/472/commits/e0f863e7748b2252697ce0801af83e3bbb83da73

File: packages/vue/src/components/organisms/SfMegaMenu/_internal/SfMegaMenuColumn.html (hunk 12 above)
Comment (pspaczek) [reply packages/vue/src/components/organisms/SfMegaMenu/_internal/SfMegaMenuColumn.html]: fixed: https://github.com/DivanteLtd/storefront-ui/pull/472/commits/6bb9a071b4e0209037779759e384dfdec4d3ad96

File: packages/vue/src/examples/pages/layout/Layout.stories.js (hunk 15 above)
Comment (pspaczek) [reply packages/vue/src/examples/pages/layout/Layout.stories.js]: fixed: https://github.com/DivanteLtd/storefront-ui/pull/472/commits/7ddf887974cc3400e66d8fd25edcb597ff9f6e4a

Comment (pspaczek) [reply packages/vue/src/examples/pages/layout/Layout.stories.js]: fixed: https://github.com/DivanteLtd/storefront-ui/pull/472/commits/7ddf887974cc3400e66d8fd25edcb597ff9f6e4a

File: packages/vue/src/components/organisms/SfMegaMenu/_internal/SfMegaMenuColumn.js (hunk 14 above)
Comment (pspaczek) [reply packages/vue/src/components/organisms/SfMegaMenu/_internal/SfMegaMenuColumn.js]: you have right, fixed: https://github.com/DivanteLtd/storefront-ui/pull/472/commits/9c97e5fdf97667972dfb156c80fdc780a0876faa

File: packages/vue/src/components/organisms/SfMegaMenu/_internal/SfMegaMenuColumn.js (hunk 13 above)
Comment (pspaczek) [reply packages/vue/src/components/organisms/SfMegaMenu/_internal/SfMegaMenuColumn.js]: fixed: https://github.com/DivanteLtd/storefront-ui/pull/472/commits/9c97e5fdf97667972dfb156c80fdc780a0876faa

File: packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.js (hunk 11 above)
Comment (pspaczek) [reply packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.js]: I split these props to `title` and `visible` now it should be most descriptive :) fix: https://github.com/DivanteLtd/storefront-ui/pull/472/commits/749eec67e11793de3d49be672cfc18808e2584d5

File: packages/vue/src/examples/pages/layout/Layout.stories.js (hunk 15 above)
Comment (pspaczek) [reply packages/vue/src/examples/pages/layout/Layout.stories.js]: include SfHeader and SfMegaMenu: https://github.com/DivanteLtd/storefront-ui/pull/472/commits/99037386f6d8d4469a48ea18902fe01480ba5e04

File: packages/vue/src/components/organisms/SfHeader/SfHeader.md (hunk 10 above)
Comment (pspaczek) [reply packages/vue/src/components/organisms/SfHeader/SfHeader.md]: back here: https://github.com/DivanteLtd/storefront-ui/pull/472/commits/821c9ae64f1567525efecfd09b85e17acd6a48e8

**Comment by filrak** [comment]
Visual fix -
image
Width if image should be the same as width of navigation bar

**Comment by filrak** [comment]
Also one story is not working: 
image

**Comment by filrak** [comment]
https://deploy-preview-472--storefrontui-storybook.netlify.com/?path=/story/organisms-megamenu--slot-aside
On mobile view bottom part is cutted. Should be full-height
image

File: packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.stories.js [hunk 16]
```diff
@@ -1,8 +1,295 @@
...
+        <SfMegaMenuColumn 
+          v-for="subcategory in subcategories" 
+          :key="subcategory.header" 
+          :title="subcategory.header"
+        >
+          <SfMenuItem 
+            v-for="item in subcategory.items" 
+            :key="item.label" 
+            :label="item.label" 
+            class="sf-mega-menu__menu-item"
```
Comment (filrak) [inline packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.stories.js]: why do we need this class?

Comment (pspaczek) [reply packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.stories.js]: just for adjust to the design, we can add SfMegaMenuItem and add this class inside

Comment (pspaczek) [reply packages/vue/src/components/organisms/SfMegaMenu/SfMegaMenu.stories.js]: @filrak I added SfMegaMenuItem to wrap SfMnuItem and add class

**Comment by pspaczek** [comment]
@pspaczek add a few improvements and adjustment

**Pull request merged**